FIX: Allow default UI actions to work without setting up a project-wide actions file (ISXB-811).

graham-huws opened the pull request:
### Description

Regression from previous versions: Default UI interactions like clicking, dragging etc should work without users having to set up their own PWA file. [ISXB-811](https://jira.unity3d.com/browse/ISXB-811). Lengthy discussion in Slack [here](https://unity.slack.com/archives/C04RDRXRJ1L/p1711549077482879).

### Changes made

Load DefaultInputActions file if a project-wide action has not been set by the user. Enable the UI functions from it.
Also, fix newlines not functioning correctly in a uxml file - details [here](https://unity.slack.com/archives/C3414V4UV/p1673359236696379?thread_ts=1673347662.674689&cid=C3414V4UV).

### Checklist

Before review:

- [x] Changelog entry added.
    - Explains the change in `Changed`, `Fixed`, `Added` sections.
    - For API change contains an example snippet and/or migration example.
    - FogBugz ticket attached, example `([case %number%](https://issuetracker.unity3d.com/issues/...))`.
    - FogBugz is marked as "Resolved" with *next* release version correctly set.
- [ ] Tests added/changed, if applicable.
    - Functional tests `Area_CanDoX`, `Area_CanDoX_EvenIfYIsTheCase`, `Area_WhenIDoX_AndYHappens_ThisIsTheResult`.
    - Performance tests.
    - Integration tests.
- [ ] Docs for new/changed API's.
    - Xmldoc cross references are set correctly.
    - Added explanation how the API works.
    - Usage code examples added.
    - The manual is updated, if needed.

During merge:

- [x] Commit message for squash-merge is prefixed with one of the list:
    - `NEW: ___`.
    - `FIX: ___`.
    - `DOCS: ___`.
    - `CHANGE: ___`.
    - `RELEASE: 1.1.0-preview.3`.

graham-huws committed:
FIX: Allow default UI actions to work without setting up a project-wide actions file (ISXB-833).

Also, fix newlines not functioning correctly in uxml file.

jfreire-unity reviewed (changes requested):
Code looks ok to me 👍
But I would like see at least a small test to make sure the problem is solved. Essentially, something that shows that these default actions are working if there is no project wide actions asset available.

jfreire-unity commented:
As discussed with @graham-huws , I'll add this test while he focus on a template change fix.

jfreire-unity committed:
Evaluate InputSystemProvider has an asset

This makes sure that InputForUI input will not break during tests.

jfreire-unity committed:
Add tests which use and remove project-wide actions

Evaluates that there is an action asset fallback in case there are no project-wide actions.

unity-cla-assistant commented:
[![CLA assistant check](https://cla-assistant.cds.internal.unity3d.com/pull/badge/signed)](https://cla-assistant.cds.internal.unity3d.com/Unity-Technologies/InputSystem?pullRequest=1899) <br/>All committers have signed the CLA.

jfreire-unity committed:
Fix formatting issues

ekcoh commented on Assets/Tests/InputSystem/Plugins/InputForUITests.cs:
```diff
@@ -31,13 +31,12 @@ public override void Setup()
         m_InputSystemProvider = new InputSystemProvider();
         EventProvider.SetMockProvider(m_InputSystemProvider);
+
+        // Test assumes a compatible action asset configuration exists for UI
```
I would suggest to not have asserts in Setup and only have asserts to actually verify functional requirements as part of test methods. For what scenario would this fail were it wouldn't fail the functional tests anyway? Otherwise if it renders the test non-applicable use assumption failure or throw exception if programmer error?

graham-huws-unity replied:
I've moved this check out to a separate test in 404f0a2b9aefb2d21e7e69038f0103a5f5167189.

Pauliusd01 commented:
I'm getting this error when ProjectWideActions asset is used but does not contain the UI map:

```
NullReferenceException: Object reference not set to an instance of an object
UnityEngine.InputSystem.Plugins.InputForUI.InputSystemProvider.ReadCurrentNavigationMoveVector () (at ./Packages/com.unity.inputsystem/InputSystem/Plugins/InputForUI/InputSystemProvider.cs:215)
UnityEngine.InputSystem.Plugins.InputForUI.InputSystemProvider.DirectionNavigation (Unity.IntegerTime.DiscreteTime currentTime) (at ./Packages/com.unity.inputsystem/InputSystem/Plugins/InputForUI/InputSystemProvider.cs:164)
UnityEngine.InputSystem.Plugins.InputForUI.InputSystemProvider.Update () (at ./Packages/com.unity.inputsystem/InputSystem/Plugins/InputForUI/InputSystemProvider.cs:127)
UnityEngine.InputForUI.EventProvider.NotifyUpdate () (at D:/Gitrepo/unity/Modules/InputForUI/Provider/EventProvider.cs:164)
```
Previously I would get a warning telling me that there will be issues unless UI map is restored. Is this expected?

Pauliusd01 commented:
I am also getting this error after resetting the ProjectWideActions asset from having no UI map to its defaults:

```
InvalidOperationException: Action 'UI/Point[/Mouse/position]' must be part of an InputActionAsset in order to be able to create an InputActionReference for it
UnityEngine.InputSystem.InputActionReference.Set (UnityEngine.InputSystem.InputAction action) (at ./Packages/com.unity.inputsystem/InputSystem/Actions/InputActionReference.cs:92)
UnityEngine.InputSystem.InputActionReference.Create (UnityEngine.InputSystem.InputAction action) (at ./Packages/com.unity.inputsystem/InputSystem/Actions/InputActionReference.cs:198)
UnityEngine.InputSystem.Plugins.InputForUI.InputSystemProvider.RegisterActions (UnityEngine.InputSystem.Plugins.InputForUI.InputSystemProvider+Configuration cfg) (at ./Packages/com.unity.inputsystem/InputSystem/Plugins/InputForUI/InputSystemProvider.cs:582)
UnityEngine.InputSystem.Plugins.InputForUI.InputSystemProvider.Initialize () (at ./Packages/com.unity.inputsystem/InputSystem/Plugins/InputForUI/InputSystemProvider.cs:84)
UnityEngine.InputForUI.EventProvider.Initialize () (at D:/Gitrepo/unity/Modules/InputForUI/Provider/EventProvider.cs:125)
UnityEngine.InputForUI.EventProvider.SetEnabled (System.Boolean enable) (at D:/Gitrepo/unity/Modules/InputForUI/Provider/EventProvider.cs:60)
UnityEngine.UIElements.DefaultEventSystem.InitInputProcessor () (at D:/Gitrepo/unity/Modules/UIElements/Core/DefaultEventSystem.cs:163)
UnityEngine.UIElements.DefaultEventSystem.set_isInputReady (System.Boolean value) (at D:/Gitrepo/unity/Modules/UIElements/Core/DefaultEventSystem.cs:94)
UnityEngine.UIElements.UIElementsRuntimeUtility.RegisterPlayerloopCallback () (at D:/Gitrepo/unity/Modules/UIElements/Core/UIElementsRuntimeUtility.cs:248)
UnityEngine.UIElements.UIElementsRuntimeUtility.RegisterCachedPanelInternal (System.Int32 instanceID, UnityEngine.UIElements.IPanel panel) (at D:/Gitrepo/unity/Modules/UIElements/Core/UIElementsRuntimeUtility.cs:61)
UnityEngine.UIElements.UIElementsRuntimeUtility.FindOrCreateRuntimePanel (UnityEngine.ScriptableObject ownerObject, UnityEngine.UIElements.UIElementsRuntimeUtility+CreateRuntimePanelDelegate createDelegate) (at D:/Gitrepo/unity/Modules/UIElements/Core/UIElementsRuntimeUtility.cs:37)
UnityEngine.UIElements.PanelSettings+RuntimePanelAccess.CreateRelatedRuntimePanel () (at D:/Gitrepo/unity/Modules/UIElements/Core/GameObjects/PanelSettings.cs:160)
UnityEngine.UIElements.PanelSettings+RuntimePanelAccess.get_panel () (at D:/Gitrepo/unity/Modules/UIElements/Core/GameObjects/PanelSettings.cs:91)
UnityEngine.UIElements.PanelSettings.get_panel () (at D:/Gitrepo/unity/Modules/UIElements/Core/GameObjects/PanelSettings.cs:544)
UnityEngine.UIElements.UIDocument.SetupVisualTreeAssetTracker () (at D:/Gitrepo/unity/Modules/UIElements/Core/GameObjects/UIDocument.cs:823)
UnityEngine.UIElements.UIDocument.RecreateUI () (at D:/Gitrepo/unity/Modules/UIElements/Core/GameObjects/UIDocument.cs:619)
UnityEngine.UIElements.UIDocument.set_visualTreeAsset (UnityEngine.UIElements.VisualTreeAsset value) (at D:/Gitrepo/unity/Modules/UIElements/Core/GameObjects/UIDocument.cs:227)
UnityEngine.UIElements.UIDocument.OnValidate () (at D:/Gitrepo/unity/Modules/UIElements/Core/GameObjects/UIDocument.cs:912)
```

ekcoh commented on Assets/Tests/InputSystem/Plugins/InputForUITests.cs:
```diff
@@ -140,6 +144,37 @@ public void SendWheelEvent()
+            // Remove the project-wide actions asset in play mode and player.
+            // It will call InputSystem.onActionChange and re-set InputSystemProvider.actionAsset
+            // This the case where no project-wide actions asset is available in the project.
+            InputSystem.s_Manager.actions = null;
```
I see this bypasses the play mode check by accessing internals directly, seems like a pragmatic solution that eliminates the need to mock via IInputRuntime interface discussed previously. Just maybe note that this is not representative of what happens in reality since this can only can happen in edit mode.

I assume the intention here is to test same behavior with and without PWA. However, this should be applicable for all tests in this class, and hence I would recommend either parameterising all test cases in this class or use two separate test classes that inherits the test methods and use different fixed setups instead?

benoitalain replied:
Also we need to not corrupt the state of `InputSystem.s_Manager.actions` after running this test. I would recommend either storing its value locally and setting it back in a `finally` block, or making the state be reset in `TearDown`.

ekcoh replied:
@benoitalain This was also my first thought, but I believe its not corrupted, its automatically reinitialised as part of InputTestFixture reinitialisation.

Pauliusd01 commented:
> I'm getting this error when ProjectWideActions asset is used but does not contain the UI map:
> 
> ```
> NullReferenceException: Object reference not set to an instance of an object
> UnityEngine.InputSystem.Plugins.InputForUI.InputSystemProvider.ReadCurrentNavigationMoveVector () (at ./Packages/com.unity.inputsystem/InputSystem/Plugins/InputForUI/InputSystemProvider.cs:215)
> UnityEngine.InputSystem.Plugins.InputForUI.InputSystemProvider.DirectionNavigation (Unity.IntegerTime.DiscreteTime currentTime) (at ./Packages/com.unity.inputsystem/InputSystem/Plugins/InputForUI/InputSystemProvider.cs:164)
> UnityEngine.InputSystem.Plugins.InputForUI.InputSystemProvider.Update () (at ./Packages/com.unity.inputsystem/InputSystem/Plugins/InputForUI/InputSystemProvider.cs:127)
> UnityEngine.InputForUI.EventProvider.NotifyUpdate () (at D:/Gitrepo/unity/Modules/InputForUI/Provider/EventProvider.cs:164)
> ```
> 
> Previously I would get a warning telling me that there will be issues unless UI map is restored. Is this expected?



> I am also getting this error after resetting the ProjectWideActions asset from having no UI map to its defaults:
> 
> ```
> InvalidOperationException: Action 'UI/Point[/Mouse/position]' must be part of an InputActionAsset in order to be able to create an InputActionReference for it
> UnityEngine.InputSystem.InputActionReference.Set (UnityEngine.InputSystem.InputAction action) (at ./Packages/com.unity.inputsystem/InputSystem/Actions/InputActionReference.cs:92)
> UnityEngine.InputSystem.InputActionReference.Create (UnityEngine.InputSystem.InputAction action) (at ./Packages/com.unity.inputsystem/InputSystem/Actions/InputActionReference.cs:198)
> UnityEngine.InputSystem.Plugins.InputForUI.InputSystemProvider.RegisterActions (UnityEngine.InputSystem.Plugins.InputForUI.InputSystemProvider+Configuration cfg) (at ./Packages/com.unity.inputsystem/InputSystem/Plugins/InputForUI/InputSystemProvider.cs:582)
> UnityEngine.InputSystem.Plugins.InputForUI.InputSystemProvider.Initialize () (at ./Packages/com.unity.inputsystem/InputSystem/Plugins/InputForUI/InputSystemProvider.cs:84)
> UnityEngine.InputForUI.EventProvider.Initialize () (at D:/Gitrepo/unity/Modules/InputForUI/Provider/EventProvider.cs:125)
> UnityEngine.InputForUI.EventProvider.SetEnabled (System.Boolean enable) (at D:/Gitrepo/unity/Modules/InputForUI/Provider/EventProvider.cs:60)
> UnityEngine.UIElements.DefaultEventSystem.InitInputProcessor () (at D:/Gitrepo/unity/Modules/UIElements/Core/DefaultEventSystem.cs:163)
> UnityEngine.UIElements.DefaultEventSystem.set_isInputReady (System.Boolean value) (at D:/Gitrepo/unity/Modules/UIElements/Core/DefaultEventSystem.cs:94)
> UnityEngine.UIElements.UIElementsRuntimeUtility.RegisterPlayerloopCallback () (at D:/Gitrepo/unity/Modules/UIElements/Core/UIElementsRuntimeUtility.cs:248)
> UnityEngine.UIElements.UIElementsRuntimeUtility.RegisterCachedPanelInternal (System.Int32 instanceID, UnityEngine.UIElements.IPanel panel) (at D:/Gitrepo/unity/Modules/UIElements/Core/UIElementsRuntimeUtility.cs:61)
> UnityEngine.UIElements.UIElementsRuntimeUtility.FindOrCreateRuntimePanel (UnityEngine.ScriptableObject ownerObject, UnityEngine.UIElements.UIElementsRuntimeUtility+CreateRuntimePanelDelegate createDelegate) (at D:/Gitrepo/unity/Modules/UIElements/Core/UIElementsRuntimeUtility.cs:37)
> UnityEngine.UIElements.PanelSettings+RuntimePanelAccess.CreateRelatedRuntimePanel () (at D:/Gitrepo/unity/Modules/UIElements/Core/GameObjects/PanelSettings.cs:160)
> UnityEngine.UIElements.PanelSettings+RuntimePanelAccess.get_panel () (at D:/Gitrepo/unity/Modules/UIElements/Core/GameObjects/PanelSettings.cs:91)
> UnityEngine.UIElements.PanelSettings.get_panel () (at D:/Gitrepo/unity/Modules/UIElements/Core/GameObjects/PanelSettings.cs:544)
> UnityEngine.UIElements.UIDocument.SetupVisualTreeAssetTracker () (at D:/Gitrepo/unity/Modules/UIElements/Core/GameObjects/UIDocument.cs:823)
> UnityEngine.UIElements.UIDocument.RecreateUI () (at D:/Gitrepo/unity/Modules/UIElements/Core/GameObjects/UIDocument.cs:619)
> UnityEngine.UIElements.UIDocument.set_visualTreeAsset (UnityEngine.UIElements.VisualTreeAsset value) (at D:/Gitrepo/unity/Modules/UIElements/Core/GameObjects/UIDocument.cs:227)
> UnityEngine.UIElements.UIDocument.OnValidate () (at D:/Gitrepo/unity/Modules/UIElements/Core/GameObjects/UIDocument.cs:912)
> ```

Repro'd both on develop, disregard them. (They are filed separately: https://jira.unity3d.com/browse/ISX-1953 / https://jira.unity3d.com/browse/ISX-1954 )

ekcoh commented on Packages/com.unity.inputsystem/CHANGELOG.md:
```diff
@@ -11,6 +11,7 @@ however, it has to be formatted properly to pass verification tests.
 ## [Unreleased] - YYYY-MM-DD
 
 ### Fixed
+- Default UI actions would not function without a user specifically setting up project-wide actions in Project Settings [ISXB-811](https://issuetracker.unity3d.com/product/unity/issues/guid/ISXB-811).
```
I would recommend rewriting this in the format that explains the problem+symptom and new behavior. E.g. (not sure if below is 100% correct, but just to illustrate what I mean)

Fixes an issue where UGUI or UI Toolkit would not react to input unless a Project-wide Input Action Asset was configured for the project. If no Project-wide Input Action Asset has been configured for the project, default actions are now instead created on-the-fly to provide default bindings for UI actions.

graham-huws-unity replied:
Tweaked this in 982b8abdd43d4e7f05fa560d305f2f5f972c6fef.

ekcoh commented on Packages/com.unity.inputsystem/InputSystem/Editor/UITKAssetEditor/Resources/InputActionsProjectSettings.uxml:
```diff
@@ -8,7 +8,7 @@
     </ui:VisualElement>
     <ui:VisualElement name="missing-asset-section">
-        <ui:HelpBox text="Actions for the Input System are stored in an Action Asset. You can assign an Action Asset as project-wide in the field above to make them accessible directly through the InputSystem.actions API.\n\nClick the button below to create a new Action Asset containing default actions, which will be assigned as project-wide." message-type="Info"/>
+        <ui:HelpBox text="Actions for the Input System are stored in an Action Asset. You can assign an Action Asset as project-wide in the field above to make them accessible directly through the InputSystem.actions API.&#10;&#10;Click the button below to create a new Action Asset containing default actions, which will be assigned as project-wide." message-type="Info"/>
```
Is this an additional fix not called out or a reformatting mistake?

graham-huws-unity replied:
This is called out in "Changes made" in the PR description.

ekcoh replied:
Sorry, guess I overlooked it

ekcoh commented on Packages/com.unity.inputsystem/InputSystem/Plugins/InputForUI/InputSystemProvider.cs:
```diff
@@ -16,7 +16,9 @@ internal class InputSystemProvider : IEventProviderImpl
 
         Configuration m_Cfg;
 
+        public InputActionAsset inputActionAsset => m_InputActionAsset;
```
Why did this change to public visibility?

graham-huws-unity replied:
Updated in e99ed66a5b8bc1887f3527cf35d03b809e34d548 to be more restrictive.

ekcoh reviewed (commented):
Thanks for submitting this PR. Fix looks good since it avoids regressive behavior even if it might imply differentiation between PWA and DefaultActions bindings. 

I still think this needs more test coverage. I would like to see that all UI actions have associated test cases (may happen outside this PR but would not suggest delaying). We may not necessarily cover all bindings (but it would be good for defaults) but if we do not cover all actions this may break without us noticing. Hence I would recommend adding test coverage for all actions.

In addition it would be good to add tests that covers:
- Behavior when assigned PWA asset is modified after being assigned to InputSystemProvider. 
- Behavior when assigned PWA asset is not containing required actions at all.
- Behavior when assigned PWA asset is partially containing required actions.
- Behavior when assigned PWA asset is containing required actions but with invalid action types.
- Behavior when assigned PWA asset is containing required actions but has no bindings on them.

I am not sure if this PR is intended to cover the task of improving the UI integration test coverage but if it is I would recommend covering tests above. If desired to handle outside this PR to land fix this is OK but then test coverage shouldn't be postponed. Feel free to reach out directly for further discussion.

ekcoh reviewed (changes requested):
See previous comments on test coverage (seem to also have problems in the area found by Paulius). I am open to handle parts of the test coverage outside this PR if it reduces delay, but do not suggest postponing it. Also have some open questions.

Pauliusd01 reviewed (approved):
LGTM. Checked UITK functionality with ProjectWideActions asset and without, in different backend modes (Old, New, Both) and with/without EventSystem object - works fine now.

I do think that https://jira.unity3d.com/browse/ISX-1953 / https://jira.unity3d.com/browse/ISX-1954 need to be fixed for it to be in a good state though but they are not caused by this PR.

benoitalain commented on Packages/com.unity.inputsystem/InputSystem/Plugins/InputForUI/InputSystemProvider.cs:
```diff
@@ -577,17 +579,14 @@ void RegisterActions(Configuration cfg)
-                m_RightClickAction = InputActionReference.Create(m_InputActionAsset.FindAction(m_Cfg.RightClickAction));
-                m_ScrollWheelAction = InputActionReference.Create(m_InputActionAsset.FindAction(m_Cfg.ScrollWheelAction));
-            }
+            m_PointAction = InputActionReference.Create(m_InputActionAsset.FindAction(m_Cfg.PointAction));
```
We're passing `cfg` as an argument but we're using `m_Cfg`. Either don't pass `cfg` at all or use it instead of the global variable.

graham-huws-unity replied:
Fix in 947912723ac498e31c36c4f0c63c04a5c1712b9e.

benoitalain commented on Packages/com.unity.inputsystem/InputSystem/Plugins/InputForUI/InputSystemProvider.cs:
```diff
@@ -610,17 +609,22 @@ void RegisterActions(Configuration cfg)
+            if (InputSystem.actions == null)
+            {
+                // If we've not loaded a user-created set of actions, just enable the UI actions from our defaults.
+                m_InputActionAsset.FindActionMap("UI", true).Enable();
```
OK. Maybe we should only create the UI actions, instead of creating all and leaving the rest disabled? But I understand that this would make the code more error-prone and isn't necessary at all for the current fix.

graham-huws-unity replied:
Yeah, that was my line of thinking - plus, this will need a fuller re-write in the future anyway to remove DefaultInputActions.

benoitalain reviewed (approved):
I tested the fix and it works.

benoitalain commented on Assets/Tests/InputSystem/Plugins/InputForUITests.cs:
```diff
@@ -140,6 +144,37 @@ public void SendWheelEvent()
 
+    [Test]
+    [Category("InputForUI")]
+    [TestCase(true)]
```
I have this one failing locally right now.

graham-huws-unity replied:
Have been discussing with @Pauliusd01 and @ekcoh - There's an issue with the way input tests are setup which makes their behaviour inconsistent, which needs addressing as a separate issue. For example, running the whole of the "Tests.dll" playmode section makes this test pass, but running just the sub-group "InputForUITests" has this test - and on another run, the rest in that group - fail. I don't think it's worth holding this PR up, as all manual testing shows the change is working as intended.

graham-huws committed:
Merge branch 'develop' into ISXB-811-default-actions-unset

graham-huws committed:
Merge branch 'develop' into ISXB-811-default-actions-unset

graham-huws committed:
Don't need to pass around a member variable.

graham-huws committed:
Restrict visibility of private member variable to only share what's required.

graham-huws committed:
Update CHANGELOG.md

graham-huws committed:
Move assertion to separate test.

graham-huws committed:
Merge branch 'develop' into ISXB-811-default-actions-unset
